src/sinks/influxdb_metrics.rs

```diff
@@ -0,0 +1,916 @@
+        .collect()
+}
+
+fn encode_distribution(values: &[f64], counts: &[u32]) -> Option<HashMap<String, Field>> {
+    if values.len() != counts.len() {
+        return None;
+    }
+
+    let mut samples = Vec::new();
+    for (v, c) in values.iter().zip(counts.iter()) {
+        for _ in 0..*c {
+            samples.push(*v);
+        }
+    }
+
+    if samples.is_empty() {
+        return None;
+    }
+
+    if samples.len() == 1 {
+        let val = samples[0];
+        return Some(
+            vec![
+                ("min".to_owned(), Field::Float(val)),
+                ("max".to_owned(), Field::Float(val)),
+                ("median".to_owned(), Field::Float(val)),
+                ("avg".to_owned(), Field::Float(val)),
+                ("sum".to_owned(), Field::Float(val)),
+                ("count".to_owned(), Field::Float(1.0)),
+                ("quantile_0.95".to_owned(), Field::Float(val)),
+            ]
+            .into_iter()
+            .collect(),
+        );
+    }
+
+    samples.sort_by(|a, b| a.partial_cmp(b).unwrap_or(Ordering::Equal));
+
+    let length = samples.len() as f64;
+    let min = samples.first().unwrap();
+    let max = samples.last().unwrap();
+
+    let p50 = samples[(0.50 * length - 1.0).round() as usize];
+    let p95 = samples[(0.95 * length - 1.0).round() as usize];
+
+    let sum = samples.iter().sum();
+    let avg = sum / length;
+
+    let fields: HashMap<String, Field> = vec![
+        ("min".to_owned(), Field::Float(*min)),
+        ("max".to_owned(), Field::Float(*max)),
+        ("median".to_owned(), Field::Float(p50)),
+        ("avg".to_owned(), Field::Float(avg)),
+        ("sum".to_owned(), Field::Float(sum)),
+        ("count".to_owned(), Field::Float(length)),
+        ("quantile_0.95".to_owned(), Field::Float(p95)),
+    ]
+    .into_iter()
+    .collect();
+
+    Some(fields)
+}
+
+// https://v2.docs.influxdata.com/v2.0/reference/syntax/line-protocol/
+fn influx_line_protocol(
+    measurement: String,
+    metric_type: &str,
+    tags: Option<HashMap<String, String>>,
+    fields: Option<HashMap<String, Field>>,
+    timestamp: i64,
+) -> String {
+    let mut line_protocol = vec![encode_key(measurement)];
+
+    // Tags
+    let mut unwrapped_tags = tags.unwrap_or(HashMap::new());
+    unwrapped_tags.insert("metric_type".to_owned(), metric_type.to_owned());
+    line_protocol.push(format!(",{}", encode_tags(unwrapped_tags)));
+
+    // Fields
+    let unwrapped_fields = fields.unwrap_or(HashMap::new());
+    let encoded_fields = encode_fields(unwrapped_fields);
+    if encoded_fields.is_empty() {
+        return "".to_owned();
+    }
+    line_protocol.push(format!(" {}", encoded_fields));
+
+    // Timestamp
+    line_protocol.push(format!(" {}", timestamp));
+
+    line_protocol.join("")
+}
+
+fn encode_key(key: String) -> String {
+    key.replace("\\", "\\\\")
+        .replace(",", "\\,")
+        .replace(" ", "\\ ")
+        .replace("=", "\\=")
```

Comment:
It seems like `encode_key` function used not only on keys, but also on values. Does it make sense to change the name to something like `escape`?

Also, each call to `replace` on `String` causes creation of a new string and copying the data into it. I think it could have been much more performant to implement `encode_key` function similarly to this:

```rust
fn encode_key(key: String, output: &mut String) {
    for c in key.chars() {
        if "\\, =".contains(c) {
            output.push('\\');
        }
        output.push(c);
    }
}
```

Then, for example in `encode_key`, it could have been called like this:

```rust
let mut output = String::new();
encode_key(key, &mut output);
output.push('=');
encode_key(value, &mut output);
```

And so on. Ideally, the string copying should happen either if it is unavoidable or if it is invoked only at the step of sink initialization. So all functions such as `encode_tag`, `encode_timestamp`, and even `influx_line_protocol` could have been changed to take a mutable reference to `String` as a second argument and append data to it instead of creating a new `String` instance.

Reply:
Done